File: General-DSA-Questions/Tree/8_maximumPathSum.cpp
// Maximum Sum Path in Binary Tree
// You are given a binary tree (nodes can have positive or negative values). A path means:
// You can start at any node
// You can end at any node
// You can move only parent → child connections
// Path must be continuous
// You cannot reuse a node
// Your task:
// 👉 Find the maximum possible sum of values of any such path.

// Chatgpt Notes : https://chatgpt.com/share/695fcde7-7518-8006-a98a-f16fc68c038e
// TUF Notes : https://takeuforward.org/data-structure/maximum-sum-path-in-binary-tree

// Example : Tree considered here is :
//       -10
//       /  \
//      9   20
//          / \
//         15  7
// Maximum Path Sum = 15 + 20 + 7 = 42

// Intiution:
// 🔍 Key Observation
// At any node, 4 types of paths are possible:
//        X
//       / \
//      L   R
// Possible paths using X:
// 1.  X alone
// 2. X + left subtree
// 3. X + right subtree
// 4. Left + X + Right ← (this is a "full turn" path)

// But ⚠️ Only one side can be passed upward to parent,
// because parent can continue path only in one direction.
// So:
// Case 4 can be the final answer, but cannot be passed upward.
// Only case 1,2,3 are returned upward.
// For every node Compute:
// 1. Best downward path starting from this node.
// 2. Also update the global maximum considering full turn path.
// ✨ Important Trick
// If a child gives negative contribution → ignore it (use 0 instead)
// Because negative reduces the sum.
// 🔄 What We Compute at Each Node. 
// Let:
//     left = max(0, best downward path from left)
//     right = max(0, best downward path from right)

// Then:
//     Best path passing through this node
//     left + node->val + right
// Update global answer with this.
// node->val + max(left, right)
// Return this value upward.

#include <iostream>
#include <climits>
using namespace std;

// Definition for a binary tree node.
class TreeNode {
  public:
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int maxSum = INT_MIN;   // Stores final answer

    // Returns maximum downward path sum starting from this node
    int dfs(TreeNode* root) {
        // Base Case
        if(root == NULL) return 0;

        // Step 1: Get best contributions from left & right
        int left = max(0, dfs(root->left)); // Ignore negative contributions, use 0 instead, calculate best downward path from left
        int right = max(0, dfs(root->right)); // Ignore negative contributions, use 0 instead, calculate best downward path from right

        // Step 2: Best path passing through current node
        int currentPath = left + root->val + right;

        // Step 3: Update global answer
        maxSum = max(maxSum, currentPath);

        // Step 4: Return best single-side path to parent
        return root->val + max(left, right);
    }

    int maxPathSum(TreeNode* root) {
        dfs(root);
        return maxSum;
    }
};

// Function to create a sample tree
TreeNode* createSampleTree() {
    TreeNode* root = new TreeNode(-10);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    return root;
}

int main() {
    TreeNode* root = createSampleTree();
    Solution sol;
    cout << "Maximum Path Sum: " << sol.maxPathSum(root) << endl; // Output: 42
    return 0;
}

File: JS-DSA-Questions/Tree/8_maximumPathSum.js
// Maximum Path Sum in a Binary Tree.
// A path may start and end at any nodes, moving only along parent↔child edges, using each node once.
// Find the maximum possible sum of node values along any such path.

// TUF Notes : https://takeuforward.org/data-structure/maximum-sum-path-in-binary-tree

// Example:
//       -10
//       /  \
//      9   20
//          / \
//         15  7
// Maximum Path Sum = 15 + 20 + 7 = 42

// Key idea (DFS):
// At each node, the best path that "turns" through it is left + node.val + right (update the global
// answer with this). But only ONE side can be extended upward to the parent, so we RETURN
// node.val + max(left, right). Negative child contributions are clamped to 0 (ignored).

// JavaScript note: C++ used a class member `maxSum` initialized to INT_MIN. Here we carry it in a
// closure variable initialized to -Infinity.
class TreeNode {
  constructor(x) {
    this.val = x;
    this.left = null;
    this.right = null;
  }
}

function maxPathSum(root) {
  let maxSum = -Infinity;

  function dfs(node) {
    if (node === null) return 0;

    const left = Math.max(0, dfs(node.left)); // clamp negatives to 0
    const right = Math.max(0, dfs(node.right));

    // Best path that turns through this node
    const currentPath = left + node.val + right;
    maxSum = Math.max(maxSum, currentPath);

    // Return the best single-side path to the parent
    return node.val + Math.max(left, right);
  }

  dfs(root);
  return maxSum;
}

function createSampleTree() {
  const root = new TreeNode(-10);
  root.left = new TreeNode(9);
  root.right = new TreeNode(20);
  root.right.left = new TreeNode(15);
  root.right.right = new TreeNode(7);
  return root;
}

console.log("Maximum Path Sum:", maxPathSum(createSampleTree())); // 42
